### PR TITLE
[RLlib] Trainer sub-class for APPO (instead of using `build_trainer()`).

### PR DESCRIPTION
**Pull request by sven1977**
Trainer sub-class for APPO (instead of using `build_trainer()`).
NOTE: We inherit APPOTrainer here directly from ImpalaTrainer (which is still built via `build_trainer()`). This shows nicely that the two APIs are compatible (despite the fact that we'll soon start deprecating `build_trainer`).

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/tests/test_trainer.py [hunk 1]
```diff
@@ -33,16 +33,21 @@ def test_validate_config_idempotent(self):
         """
         # Given:
         standard_config = copy.deepcopy(COMMON_CONFIG)
+        trainer = pg.PGTrainer(env="CartPole-v0", config=standard_config)
```
Comment (gjoliver) [inline rllib/agents/tests/test_trainer.py:36]: this doesn't seem to be belong to this PR?

File: rllib/agents/trainer.py [hunk 2]
```diff
@@ -1847,9 +1847,20 @@ def resolve_tf_settings():
...
-    @staticmethod
-    def _validate_config(config: PartialTrainerConfigDict,
-                         trainer_obj_or_none: Optional["Trainer"] = None):
+    @ExperimentalAPI
+    def validate_config(self, config: PartialTrainerConfigDict) -> None:
+        """Validates a given config dict for this Trainer.
+
+        Users should override this method to implement custom validation
+        behavior. It is recommended to call `super().validate_config()` in
+        this override.
```
Comment (gjoliver) [inline rllib/agents/trainer.py:1856]: wait, we are not calling super().validate_config() in this function?

File: rllib/agents/trainer_template.py [hunk 3]
```diff
@@ -139,27 +139,34 @@ def _init(self, config: TrainerConfigDict,
...
             if validate_config is not None:
                 validate_config(config)
 
+        @staticmethod
+        @override(Trainer)
+        def execution_plan(workers, config, **kwargs):
+            # `execution_plan` is provided, use it inside
+            # `self.execution_plan()`.
+            if execution_plan is not None:
+                return execution_plan(workers, config, **kwargs)
```
Comment (gjoliver) [inline rllib/agents/trainer_template.py:163]: wait, I don't get why we need this all of a sudden. is this really related to APPO trainer?
where does this execution_plan come from?

**Review by gjoliver — COMMENTED** [review]
awesome looking PR.
just a couple of minor questions.
thanks a lot.

File: rllib/agents/tests/test_trainer.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/tests/test_trainer.py:36]: It does, we changed the `validate_config` method from `static` to normal, so now we do require an actual Trainer object to call this method, which we do here.

File: rllib/agents/trainer.py (hunk 2 above)
Comment (sven1977) [reply rllib/agents/trainer.py:1856]: This is the base method. Super (tune.Trainable) doesn't have a `validate_config`.
When users sub-class `Trainer`, they should call the super's (Trainer's) validate config to get all the logic that this default implementation already provides.

File: rllib/agents/trainer_template.py (hunk 3 above)
Comment (sven1977) [reply rllib/agents/trainer_template.py:163]: Just keeping things backward compatible (with the soon-to-be-deprecated `build_trainer` function).

New:
- Trainer now has a method: `execution_plan` with the default implementation (the one that e.g. PGTrainer uses).
- Users can override it.
- However, it is still a static method, such that backward compatibility with `build_trainer` will not break.
- It was necessary to introduce this in this PR as APPO is the first agent that we convert to sub-class that brings its own exec plan logic (PGTrainer used the default one; all other Trainers are still using `build_trainer`).

**Comment by sven1977** [comment]
Hey @gjoliver , please take another look. I have addressed all your questions. Thanks! :)

File: rllib/agents/tests/test_trainer.py (hunk 1 above)
Comment (avnishn) [reply rllib/agents/tests/test_trainer.py:36]: Why would changing it from static to normal require that the trainer call this with the standard config?

**Review by avnishn — APPROVED** [review]

File: rllib/agents/tests/test_trainer.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/tests/test_trainer.py:36]: Sorry, I think I explained in a misleading way: Not the standard config is the problem/important here. It could be any config for this test that is a ok PG config!
But you do have to call it on an instantiated object, not just call the static function like before.

File: rllib/agents/trainer.py (hunk 2 above)
Comment (gjoliver) [reply rllib/agents/trainer.py:1856]: yeah stupid me ...

File: rllib/agents/trainer_template.py (hunk 3 above)
Comment (gjoliver) [reply rllib/agents/trainer_template.py:163]: I am still a bit confused about how this gets called.
because won't the first condition always be true? like we are in execution_plan itself, so execution_plan must be there, non-None?

**Review by gjoliver — COMMENTED** [review]
nice.
